src/inference/src/ie_core.cpp

```diff
@@ -501,12 +513,132 @@ class CoreImpl : public ie::ICore, public std::enable_shared_from_this<ie::ICore
+            unsigned int optimalBatchSize = 0;
+            if (deviceNameWithBatchSize.empty()) {
+                // batch size is not set explicitly via device name e.g. BATCH:GPU(4)
+                // let's query the optimal batch size
+                std::map<std::string, ie::Parameter> options;
+                options["MODEL_PTR"] = &network;
+                optimalBatchSize = GetCPPPluginByName(DeviceIDParser(deviceNameWithoutBatch).getDeviceName())
+                                       .get_metric(METRIC_KEY(OPTIMAL_BATCH), options)
+                                       .as<unsigned int>();
+                auto res =
+                    GetConfig(deviceNameWithoutBatch, CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS)).as<std::string>();
+                requests = PerfHintsConfig::CheckPerformanceHintRequestValue(res);
+                const auto& reqs = config_with_batch.find(CONFIG_KEY(PERFORMANCE_HINT_NUM_REQUESTS));
+                if (reqs != config_with_batch.end())
+                    requests = (unsigned int)PerfHintsConfig::CheckPerformanceHintRequestValue(reqs->second);
+                if (requests) {
+                    optimalBatchSize = std::max(1u, std::min(requests, optimalBatchSize));
+                }
+            }
+            if (optimalBatchSize > 1 || !deviceNameWithBatchSize.empty()) {
+                auto function = network.getFunction();
+                bool bDetectionOutput = false;
+                for (auto&& node : function->get_ops()) {
+                    auto isDetectionOutputParent = [](decltype(node)& nd) {
+                        for (size_t n = 0; n < nd->get_input_size(); n++) {
+                            if (!std::strcmp("DetectionOutput", nd->get_input_node_ptr(n)->get_type_info().name))
+                                return true;
+                        }
+                        return false;
+                    };
+
+                    if (!std::strcmp("DetectionOutput", node->get_type_info().name) ||
+                        (!std::strcmp("Result", node->get_type_info().name) && isDetectionOutputParent(node))) {
+                        node->get_rt_info()["affinity"] = deviceNameWithoutBatch;
+                        bDetectionOutput = true;
+                    } else {
+                        node->get_rt_info()["affinity"] = "BATCH";
+                    }
+                }
+                auto batchConfig = deviceNameWithBatchSize.empty()
+                                       ? deviceNameWithoutBatch + "(" + std::to_string(optimalBatchSize) + ")"
+                                       : deviceNameWithBatchSize;
+                if (bDetectionOutput) {
+                    deviceName = "HETERO:BATCH," + deviceNameWithoutBatch;
+                    config_with_batch[CONFIG_KEY(AUTO_BATCH)] = batchConfig;
+                } else {
+                    deviceName = "BATCH:" + batchConfig;
+                }
+            }
+        } catch (std::exception& e) {
```

Comment:
Should it be `NOT_IMPLEMENTED`?

Reply:
No, we don't want to expose ANY failures related to the Auto-Batching (and silently fallback to use the batch1 network on the device directly)